test/extensions/filters/udp/udp_proxy/udp_proxy_integration_test.cc

```diff
@@ -0,0 +1,182 @@
+  Network::Address::InstanceConstSharedPtr listener_address;
+  if (version_ == Network::Address::IpVersion::v4) {
+    // Kernel regards any 127.x.x.x as local address.
+    listener_address.reset(new Network::Address::Ipv4Instance(
+#ifndef __APPLE__
+        "127.0.0.3",
+#else
+        "127.0.0.1",
+#endif
+        port));
+  } else {
+    // IPv6 doesn't allow any non-local source address for sendmsg. And the only
+    // local address guaranteed in tests in loopback. Unfortunately, even if it's not
+    // specified, kernel will pick this address as source address. So this test
+    // only checks if IoSocketHandle::sendmsg() sets up CMSG_DATA correctly,
+    // i.e. cmsg_len is big enough when that code path is executed.
+    listener_address.reset(new Network::Address::Ipv6Instance("::1", port));
+  }
+
+  requestResponseWithListenerAddress(*listener_address);
+}
+
+TEST_P(UdpProxyIntegrationTest, MultipleClients) {
+  const uint32_t port = lookupPort("listener_0");
+  auto listener_address = Network::Utility::resolveUrl(
+      fmt::format("tcp://{}:{}", Network::Test::getLoopbackAddressUrlString(version_), port));
+
+  UdpSyncClient client1(timeSystem(), version_);
+  client1.write("client1_hello", *listener_address);
+
+  UdpSyncClient client2(timeSystem(), version_);
+  client2.write("client2_hello", *listener_address);
+  client2.write("client2_hello_2", *listener_address);
+
+  Network::UdpRecvData client1_request_datagram;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(client1_request_datagram));
+  EXPECT_EQ("client1_hello", client1_request_datagram.buffer_->toString());
+
+  Network::UdpRecvData client2_request_datagram;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(client2_request_datagram));
+  EXPECT_EQ("client2_hello", client2_request_datagram.buffer_->toString());
+  ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(client2_request_datagram));
+  EXPECT_EQ("client2_hello_2", client2_request_datagram.buffer_->toString());
+
+  // We should not be getting datagrams from the same peer.
+  EXPECT_NE(*client1_request_datagram.addresses_.peer_, *client2_request_datagram.addresses_.peer_);
+
+  // Send two datagrams back to client 2.
+  fake_upstreams_[0]->sendUdpDatagram("client2_world", *client2_request_datagram.addresses_.peer_);
+  fake_upstreams_[0]->sendUdpDatagram("client2_world_2",
+                                      *client2_request_datagram.addresses_.peer_);
+  Network::UdpRecvData response_datagram;
+  client2.recv(response_datagram);
+  EXPECT_EQ("client2_world", response_datagram.buffer_->toString());
+  client2.recv(response_datagram);
+  EXPECT_EQ("client2_world_2", response_datagram.buffer_->toString());
+
+  // Send 1 datagram back to client 1.
+  fake_upstreams_[0]->sendUdpDatagram("client1_world", *client1_request_datagram.addresses_.peer_);
+  client1.recv(response_datagram);
+  EXPECT_EQ("client1_world", response_datagram.buffer_->toString());
+}
+
```

Comment:
Can we also tests that multiple packets from the same client goes to the same upstream host?